Lazy document initialization

iwehrman opened the pull request:
This PR is in the same spirit as #2439, which delayed loading detailed information about layers until their first selection, but for documents. It changes `documents.initInactiveDocuments` to only load a small subset of essential document properties during startup, and `documents.selectDocument` to finish loading all of the properties when the document is first selected. It does _not_ change `initActiveDocument` or `updateDocument` or anything else that might initialize a new document after startup. An uninitialized document is detected simply by having a null `layers` property.

With 7 moderately sized documents open, this reduces overall startup time from ~6.5 seconds to ~4.5 seconds, which is a ~30% reduction.

Although #2439 was a fairly simple architectural change, this is even simpler. Still, these sorts of performance optimizations always introduce some risk, so I'd appreciate some help testing this.

Partially, indirectly addresses #790.

volfied commented:
Changes look good, but the layer initialization PR burned my tongue one too many times. I'm going to let this one sit for a while and think about where, if any, we request another document's properties *before* selecting/initializing it.

mcilroyc commented on src/js/actions/documents.js:
```diff
@@ -647,11 +665,14 @@ define(function (require, exports) {
+                    // The now-active document has yet to be fully initialized.
+                    return this.transfer(updateDocument);
+                } else {
+                    this.dispatch(events.document.SELECT_DOCUMENT, {
```
did you intend not to use `return this.dispatchAsync...`?  Also, in the case that you transfer to updateDocument, wouldn't we also need to emit SELECT_DOCUMENT for the application store?

iwehrman replied:
1. No: `dispatch` is perfectly safe inside `then` handlers. 
2. No: `updateDocument`, when no ID is supplied, updates the _active_ document, which updates the application store accordingly.

mcilroyc commented:
Also looks good to me.  I am worried about places throughout where we assume layers is non-null ... but I guess mostly we rely on first fetching current document, so ok?

iwehrman commented:
Well, I figured out one time I hadn't anticipated previously in which `layers` is expected to be non-null: the search bar. So, I'm gonna fix that up...

iwehrman commented:
Rebased and pushed another commit which fixes search. When the search bar is toggled open, we check for uninitialized documents and call `updateDocument` on them. This _can_ be a little slow, but... it seems worth it?

I also changed `updateDocument` to not fetch guide info for non-active documents, since a) this failed and b) it wasn't necessary anyway, because that was already happening correctly in `selectDocument`. Back to @volfied!
